### ISSUE TITLE
Resubscribing to a closed broadcast receiver will hang on a call to `recv`

### ISSUE THREAD

**Issue by nathaniel-daniel** [tokio-rs/tokio#4814]
**Version**
tokio v1.19.2, tokio master (4daeea8cad1ce8e67946bc0e17d499ab304b5ca2)

**Platform**
Windows 10 64 bit

**Description**
Attempting to resubscribe to a closed broadcast receiver will hang on calls to `recv`.

I tried this code:
`Cargo.toml`:
```toml
[package]
name = "tokio-broadcast-bug"
version = "0.0.0"
edition = "2021"

[dependencies]
tokio = { version = "1.19.2", features = ["full"] }
```
`main.rs`:
```rust
#[tokio::main]
async fn main() {
    let (tx, rx) = tokio::sync::broadcast::channel::<u32>(4);
    drop(tx);

    let mut rx_clone = rx.resubscribe();
    drop(rx);

    loop {
        match rx_clone.recv().await {
            Ok(msg) => {
                println!("{}", msg);
            }
            Err(tokio::sync::broadcast::error::RecvError::Closed) => {
                println!("Closed");
                break;
            }
            Err(tokio::sync::broadcast::error::RecvError::Lagged(n)) => {
                println!("Lagged by {n} messages");
            }
        }
    }

    println!("Done");
}
```
I expected to see this happen: 
The loop should exit.

Instead, this happened: 
The program hangs indefinitely. 

Furthermore, replacing the loop with a call to `try_recv` yields an `Empty` error instead of a `Closed` error.

**Comment by Darksonn** [comment]
That certainly does sound like a bug.

**Comment by nathaniel-daniel** [comment]
I don't really understand everything about this channel, but shouldn't [`new_receiver`](https://github.com/tokio-rs/tokio/tree/ad942de2b738c3e2b99cebb0bf71b121980de194/tokio/src/sync/broadcast.rs#L661) update the next field similarly to how `recv_ref` does so [here](https://github.com/tokio-rs/tokio/blob/ad942de2b738c3e2b99cebb0bf71b121980de194/tokio/src/sync/broadcast.rs#L836-L856), by including an adjust value to account for the close message? Making this change seems to fix my simple example, but I don't know if it works in a larger project or if it introduces any other bugs; I was hoping someone more knowledgeable could take a look.

**Comment by Darksonn** [comment]
The index not being adjusted for the close message does indeed appear to be the correct answer.

**Comment by hds** [comment]
I came up with a solution for this issue which is a bit more drastic that I originally anticipated (I'm removing `closed` flag in the slot and just relying on the channel level `closed` flag).

@nathaniel-daniel Since it was your observation that the `next` flag on the resubscribed receiver that was incorrect, perhaps you'd be interested in looking at the solution? I probably wouldn't have arrived at it without your comments!

**Comment by nathaniel-daniel** [comment]
Sorry for the rate reply, I've been busy. Thanks for the fix @hds! I just tested and it seems to work great on a simple example and a small sized project I'm working on. I was writing up a PR to do essentially what you did, but it stalled as I tried to figure out the ramifications of removing the double closed flags, performance or otherwise; I couldn't figure out why the two closed flags were even introduced in the first place. Thanks for the PR!
